transformations/multi_pivot_paraphrases_generation/README.md

```diff
@@ -0,0 +1,44 @@
+# From one English Sentence to a list of paraphrases 🦎  + ⌨️ → 🐍
+This transformation generates a list of paraphrases for an English sentence by leveraging Pivot-Transaltion approach.
+Pivot-Transaltion is an approach where a sentence in a source language is translated to a foreign language called the pivot language then translated back to the source language to get a paraprhase candidate, e.g. translate an English sentence to French, then translate back to English.
+
+The paraphrases generation is divided into two step:
+- Step 1: paraphrases Candidate Over-generation by leveraging Pivot-Transaltion. At this step, we generate a Pool of possible parparhases.
+- Step 2: apply a candidate selection over the Pool of paraphrases, since the pool can contain semantically unrelated or duplicate paraphrases.
+    We leverage Embedding Model such as Universal Sentence Encoder~(USE) to disqualify candidate paraphrases from the pool, by computing the Cosine Similarity socres of the
+    USE Embeddings between the reference sentence and the candidate paraphrase. Let R = USE_Embeding(reference_english_sentence) and P = USE_Embeding(candidate):
+    - if Cosine(R,P) < alpha => the candidate is semantically unrelated and then removed from the final list of paraphrases
+    - if Cosine(R,P) > beta => the candidate is a duplication and then removed from the final list of paraphrases
+    - By default Alpha=0.5 and Beta=0.95, we set the value as suggested by [Parikh et al.](https://arxiv.org/pdf/2004.03484.pdf) works
+
+Please refer to the test.json for all of the test cases catered.
+
+This transformation translates an English sentence to a list of predefined languages using Huggingface MariamMT and EasyNMT as Machine Transaltion models.
+- The transformation support Two Pivot-Transaltion Level.
+    - If Pivot-level = 1 => Transalte to only one foreign language. e.g. English -> French -> English  ||  English -> Arabic -> English  ||  English -> japanese -> English
+    - If Pivot-level = 2 => Transalte to only Two foreign language. e.g. English -> French -> Arabic -> English  ||  English -> Russian -> Chinese -> English
+
+Author name: Auday Berro (audayberro@gmail.com)
+
+## What type of a transformation is this?
+This transformation is a paraphrase generation for Natural English Sentences by lveraging Pivot-Transaltion techniques. The Pivot-Trnasaltion technique allow to get lexically and syntaxically diverse paraphrases.
+
```

Comment:
As @mille-s mentioned, you might want to specify how this is different from earlier PRs.

Reply:
Thank you for this question. 

In general to generate paraphrases we followed a data-flow principle, by splitting the process to 2 main step:
a) candidate over-generation: generate as many possible candidate using Pivot-translation techniques trough a **predefined set of pivot** languages 
b) candidate selection: where semantically irrelevant paraphrases candidates are removed from the finale list through cosine similarity scores.

To resume here are the differences:

1. We generate the paraphrase by pivot translation techniques using two pivot-levels(1-pivot meaning we have one pivot language or 2-pivot meaning we have 2 pivot languages) that are configurable, the user can choose from the beginning the level of paraphrase generation. e.g. 1-level => English-**Italian**-English || 2-level => English-**Chinese**-**Russian**-English.
2. What makes our work different from others is that we use a manually defined list of pivot languages so that the sentences are more distinct and semantically related to the reference sentence. The languages were selected respecting two criteria:a) more the grammar of the pivot language is different from the source language(in our case is English) more we get syntactical diversity; b) more the grammar of the pivot language is close to the source language more we get lexical diversity and semantic relatedness. To resume we don't use one pivot-language as in the other works, instead we use the entire list of predefined languages respecting the selected pivot-level. e.g. if you choose 1-pivot level the paraphrases will be generated by translating respectively to each language in the list and translate back to English.
3. Generate paraphrases is not enough we should ensure that the paraphrases candidates are semantically related to the reference sentences, since the translator Machine Engine may generate duplicate and semantically unrelated sentences. so the concept is to ensure that the result is of high quality(in our case semantically related to the reference sentences) so we need to perform a quality control step, it can be during generation or after generation of the paraphrase. In our transformation we apply quality control after paraphrase generation by computing the cosine similarity of the embedding vector off the reference sentence and the candidate paraphrase. We support Universal Sentence Encoder embedding, we can add other embedding model like BERT and ELMO but due to time constraint we used USE.
4. Candidate selection as mentioned in 3 is configurable, the user can choose to apply or not after generation.
5. The Semantic relatedness threshold are configurable and can be changed, in our work we used a minimal score of 0.5(if cosine score is lower than 0.5 the candidate is considered as semantically unrelated)